Fix: PetNamesBridge to use InternalAccounts

montelaidev opened the pull request:
## **Description**

This PR updates the PeNamesBridge to use InternalAccounts from the AccountsController instead of using identities. 

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/22933

## **Manual testing steps**

1. Create a snap account
2. Open extension and click on send
3. Select a snap account as a recipient
4. See that the name is correct. 

## **Screenshots/Recordings**

### **Before**

![Screenshot 2024-02-19 at 21 48 00](https://github.com/MetaMask/metamask-extension/assets/96463427/8f04b481-696e-47d8-b3aa-2a50bc8baee3)


### **After**

![Screenshot 2024-02-19 at 21 36 48](https://github.com/MetaMask/metamask-extension/assets/96463427/4086e626-4be5-47fa-a527-ca4c700a9780)


## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

montelaidev committed:
fix: PetNamesBridge to use InternalAccounts

dbrans commented on app/scripts/lib/AbstractPetnamesBridge.ts:
```diff
@@ -28,13 +32,14 @@ enum SyncDirection {
 
-type AllowedEvents = NameStateChange;
+type AllowedEvents = NameStateChange | AccountsControllerChangeEvent;
+type AllowedActions = AccountsControllerListAccountsAction;
```
Let's try not to mix account events and actions and actions in this file. It makes our abstract class too aware of its subclasses' details. How about we fix this by:

1. Letting AbstractPetnamesBridge work with a more generic messenger.
2. Allowing PetnamesBridgeMessenger to use template parameters for actions.
3. Using a different messenger for the subclass.

Or, is there another approach we could consider?

montelaidev replied:
Sure, we can make it work with a more generic messenger

dbrans commented on app/scripts/lib/AccountIdentitiesPetnamesBridge.test.ts:
```diff
@@ -75,48 +92,54 @@ function createNameStateWithPetname(
-  };
+function setupAccountsController(
+  messenger,
+  state = EMPTY_ACCOUNTS_CONTROLLER_STATE,
```
```suggestion
  messenger: AccountsControllerMessenger,
```

montelaidev committed:
fix: make messenger more generic

dbrans reviewed (commented):
Hi @montelaidev, nice work on the changes. I have a few suggestions, so I put together a PR which you can merge into your branch if you choose: https://github.com/MetaMask/metamask-extension/pull/23088.

The key change is not exporting PetnamesBridgeAllowedEvents/Actions and instead to pass in the events/actions that the subclasses need directly to AbstractPetnamesBridge and PetnamesBridgeMessenger.

dbrans committed:
Suggested changes to fix/22933 (#23088)

## **Description**
Suggesting changes to
https://github.com/MetaMask/metamask-extension/pull/23046.

The key change is not exporting PetnamesBridgeAllowedEvents/Actions and
instead to pass in the events/actions that the subclasses need directly
to AbstractPetnamesBridge and PetnamesBridgeMessenger.
<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/PR?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it
is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [ ] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into fix/22933

dbrans reviewed (dismissed):
Monte, looks good to me!

I'll ask @matthewwalsh0 for another pair of eyes.

matthewwalsh0 commented on app/scripts/lib/AbstractPetnamesBridge.ts:
```diff
@@ -28,14 +32,17 @@ enum SyncDirection {
 
-export type PetnamesBridgeMessenger = RestrictedControllerMessenger<
+export type PetnamesBridgeMessenger<
+  Event extends EventConstraint,
```
Depending on our TypeScript version, can we use defaults here to assign these both to `never` if not provided, so it's not required on every bridge implementation such as the address book?

matthewwalsh0 commented on app/scripts/lib/AccountIdentitiesPetnamesBridge.test.ts:
```diff
@@ -129,51 +162,80 @@ function createNameControllerMock(
+  >;
+
   beforeEach(() => {
+    messenger = new ControllerMessenger();
```
I'm a big fan of isolating unit tests as much as possible by mocking anything we don't want to test.

We're using a real instances of the controller messenger meaning if that code fails, these tests now fail also.

In this scenario, we can simulate the messenger events by provided an entirely mocked `messenger` object to the constructor and then just invoking the first argument passed to the `subscribe` method as that's the limit of our interaction in the implementation.

dbrans replied:
@montelaidev – check out the way the messenger is mocked in AddressBookPetnamesBridge.test.ts – and how the first argument, the listener callback, is invoked:
https://github.com/MetaMask/metamask-extension/blob/b16ad3171e7c2b36d611f23c43b5ebb532fdfadd/app/scripts/lib/AddressBookPetnamesBridge.test.ts#L138-L140

montelaidev replied:
That makes sense, i've changed the tests to use a mocked messenger

matthewwalsh0 commented on app/scripts/lib/AccountIdentitiesPetnamesBridge.ts:
```diff
@@ -4,44 +4,52 @@ import {
-    return Object.values(identities).map((identity: AccountIdentityEntry) => ({
-      value: identity.address,
+    const internalAccounts = this.messenger.call(
+      'AccountsController:listAccounts',
```
Just confirming for audit sake this will still include all the standard account identities, but now also the snap account names?

montelaidev replied:
Yes, the accounts controller will have all the same accounts that the preference controller has.

matthewwalsh0 commented on test/jest/mocks.js:
```diff
@@ -148,3 +152,39 @@ export const getInitialSendStateWithExistingTxState = (draftTxState) => ({
 });
+
+export function createMockInternalAccount({
+  address = '0xd5e099c71b797516c10ed0f0d895f429c2781111',
```
If the intent is to centralise these mocks to avoid duplication, is it worth putting this address in a constant such as `ACCOUNT_MOCK` so we can assert against it later?

montelaidev committed:
fix: use mock messenger

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into fix/22933

montelaidev committed:
chore: fix lint

montelaidev committed:
chore: fix typing

dbrans commented on app/scripts/lib/AbstractPetnamesBridge.ts:
```diff
@@ -35,8 +35,8 @@ type ChangeList = Record<ChangeType, PetnameEntry[]>;
 export type PetnamesBridgeMessenger<
-  Event extends EventConstraint,
-  Action extends ActionConstraint,
+  Event extends EventConstraint = PetnamesBridgeAllowedEvents,
```
You want this to be `never`. The reason being that we are already taking the union of Event with PetnamesBridgeAllowedEvents on line 43, so this default is confusing.

dbrans commented on app/scripts/lib/AbstractPetnamesBridge.ts:
```diff
@@ -64,8 +64,8 @@ function getKey({ type, variation, value }: PetnameEntry): string {
 export abstract class AbstractPetnamesBridge<
-  Event extends EventConstraint,
-  Action extends ActionConstraint,
+  Event extends EventConstraint = PetnamesBridgeAllowedEvents,
```
You want this to `= never` for the same reason above. See comment on line 38.

dbrans commented on app/scripts/lib/AccountIdentitiesPetnamesBridge.test.ts:
```diff
@@ -97,59 +90,22 @@ function createNameStateWithPetname(
-  AccountIdentitiesPetnamesBridgeEvents,
-  AccountIdentitiesPetnamesBridgeActions
+function createMessengerMock(): jest.Mocked<
+  PetnamesBridgeMessenger<
```
If you're using this type a lot it can help to define it somewhere to avoid repetition.
```
type BridgeMessenger = PetnamesBridgeMessenger<
    AccountIdentitiesPetnamesBridgeEvents,
    AccountIdentitiesPetnamesBridgeActions
  >;
```

montelaidev committed:
fix: update types for bridge

matthewwalsh0 commented on app/scripts/lib/AddressBookPetnamesBridge.ts:
```diff
@@ -11,7 +11,10 @@ import {
 
-export class AddressBookPetnamesBridge extends AbstractPetnamesBridge {
+export class AddressBookPetnamesBridge extends AbstractPetnamesBridge<
+  never,
```
Can we remove these since we added the default templates?

montelaidev replied:
Yes we can remove these

matthewwalsh0 commented on app/scripts/lib/AccountIdentitiesPetnamesBridge.test.ts:
```diff
@@ -129,24 +109,45 @@ function createNameControllerMock(
+    // mock listAccounts call
+    messenger.call.mockReturnValue([MOCK_INTERNAL_ACCOUNT]);
+
+    const listener = messenger.subscribe.mock.calls[0][1] as (
```
Minor, but as it's a bit verbose to include in each test, we'd usually make a helper function such as:

`simulateSubscribe(messenger: jest.Mocked<AccountIdentitiesPetnamesBridgeMessenger>, stateChange: AccountsControllerState, patch: any[])`

montelaidev replied:
Sure, that makes sense.

montelaidev committed:
fix: remove never from AddressBookBetnamesBridge and add helper function

dbrans commented:
Nice work!

k-g-j commented:
Works as expected in manual testing steps in Chrome Version 122.0.6261.112

https://github.com/MetaMask/metamask-extension/assets/91970214/ad1d4957-aff7-4b34-99cf-e56736b9f578

Works as expected in manual testing steps in Firefox Version 123.0.1

https://github.com/MetaMask/metamask-extension/assets/91970214/66674f2a-b1fa-4f1d-882c-b054f411020a
